### ISSUE TITLE
Unable to get Document's content using MS Graph Client

### ISSUE THREAD

**Issue by AJIXuMuK** [SharePoint/sp-dev-docs#3077]
#### Category
- [ ] Question
- [ ] Typo
- [x] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

MS Graph has an API to get file's (drive item) content: [documetation](https://docs.microsoft.com/en-us/graph/api/driveitem-get-content?view=graph-rest-1.0) and event convert it to another format.

It would be great if we can use it in SPFx solutions like:
```typescript
const graphClient = await this.context.msGraphClientFactory.getClient();
const pageContext = this.context.pageContext;
await graphClient.api(`sites/${window.location.hostname},${pageContext.site.id},${pageContext.web.id}/lists/${pageContext.list.id}/items/${itemId}/driveItem/content?format=pdf`).version('v1.0').get();
```

#### Observed Behavior
preflight request returns an error:
```
Access to fetch at 'https://northcentralus1-mediap.svc.ms/transform/pdf?provider=spo&inputFormat=docx&cs=MjIyMTg5NTYtNTU1Ni00Y2E1LWJiNTAtMmVjMTMyMTI2OTlhfFNQTw&correlationId=9cccb596-3a1e-4888-a111-dc2d0922cae5&docid=https:%2F%2Fsplst.sharepoint.com%2F_api%2Fv2.0%2Fdrives%2Fb!rqMrjkKEc0imnTlehKinAefrVZvchCtFrJgaGMr2BZgislpczV2bTpP2qEHeyD07%2Fitems%2F016ILJLEE24PKEAUJXFVA2W43G2PXLML4D%3Ftempauth%3token-goes-here%26version%3DPublished' (redirected from 'https://graph.microsoft.com/v1.0/sites/splst.sharepoint.com,site-id,web-id/lists/list-id/items/1/driveItem/content?format=pdf') from origin 'null' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```
And after that no actual request is done.
![no-requests](https://user-images.githubusercontent.com/17036219/49540754-b4a74500-f885-11e8-99c2-a1e98802a2fa.png)

If you look at Graph Explorer it does have the same preflight error but it still sends actual requests after that:
![ms-graph](https://user-images.githubusercontent.com/17036219/49540855-fa640d80-f885-11e8-82e2-21c83f0f2333.png)

#### Steps to Reproduce

Use the code above in Command Set Customizer.

**Comment by patmill** [comment]
Hi Alex, can you shed a bit more light on what you plan to do with the result?  As a co-worker says "They’re on track to loading a multi-megabyte PDF file into a JS variable".


All, what’s going on here is:
-	Browser makes AJAX call to Graph
    o	Browser sends CORS preflight to Graph and succeeds
-	Graph (we) return a 302 redirect to MeTA for the download
    o	   Browser has a choice:
        	Fail, not supporting redirects at all.
               •	lLuckily most browsers support redirects
        	Follow the redirect, providing the current domain as the ‘origin’ as per CORS
                •	There’s a tricky (but contrived) exploit that is enabled if browsers do this.
        	Follow the redirect, providing ‘null’ as the origin
                •	This is the workaround that browsers employ

The media transformation service (the call to /transform/pdf) isn't handling the null origin very well, so we're looking into that.  In the meantime, we're wondering if there is still a solution to your problem that can be solved in your code.

**Comment by AJIXuMuK** [comment]
For now I was just playing with that conversion mechanism as one of my customers is potentially interested is such a functionality.
So, it's not a blocker for now.

But the original idea is to convert the file, save it to the selected doc library.

> They’re on track to loading a multi-megabyte PDF file into a JS variable

Any other ideas on how to save the conversion results into document library?

**Comment by patmill** [comment]
Not off the top of my head, unfortunately, at least nothing straight forward.  Kicking off a call to a service that does it could work, but that is certainly a higher bar - but might be more robust.  The user navigating away, closing the browser, etc. won't kill the work.  At any rate, looking into the CORS issue still.

**Comment by lahuey** [comment]
@AJIXuMuK  are you still seeing issues with this problem?

**Comment by AJIXuMuK** [comment]
Hi @patmill 
I'm not using it right now.
But I can check it early next week to see if there are any changes on my end.

**Comment by lahuey** [comment]
@AJIXuMuK Please let us know as soon as possible. We would like to investigate the problem if you're still having issues.

Thanks!

**Comment by AJIXuMuK** [comment]
Hi @patmill,
The error is the same but in other place.
Now, both preflight and actual requests are sent.
But when browser tries to get info from `https://northcentralus1-mediap.svc.ms/transform/pdf` it still gets CORS error.
So, in the code of Command Set I still get 'Failed to fetch'.
The CORS error:
```
Access to fetch at 'https://northcentralus1-mediap.svc.ms/transform/pdf?provider=spo&inputFormat=docx&cs=Mjg0MTA1OWYtMDE5Zi00NGFmLTg3MzAtODM3NDYwN2FkOWQ5fFNQTw&correlationId=dd941502-4d91-4563-b8b4-c05ddb483409&docid=https:%2F%2Faterentiev.sharepoint.com%2F_api%2Fv2.0%2Fdrives%2Fb!vFIvV9hJ_EGKkVmW2myXa9t0u9zsPKlHonEyWhW5MDRXBrC4p-n1T4YJ8AVxwMoO%2Fitems%2F01NDOFDTN6RE6LI74BAZAI53XLOBXJF62P%3Ftempauth%3DeyJ0eXAiOiJKV1QiLCJhbGciOiJub25lIn0.eyJhdWQiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAvYXRlcmVudGlldi5zaGFyZXBvaW50LmNvbUBkNGFmMmNiOC0wYjc4LTQ5YTQtYjViZC1jODJmZWE0MjBiYzgiLCJpc3MiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAiLCJuYmYiOiIxNTU3MzM4NTc2IiwiZXhwIjoiMTU1NzM2MDE3NiIsImVuZHBvaW50dXJsIjoiWnhkUHF6TTJGdFZsdzU2QlZVL3RUUXJMRzNLSHR4VVhTUWFCaE1rdEZVST0iLCJlbmRwb2ludHVybExlbmd0aCI6IjE1OCIsImlzbG9vcGJhY2siOiJUcnVlIiwiY2lkIjoiWkdRNU5ERTFNREl0TkdRNU1TMDBOVFl6TFdJNFlqUXRZekExWkdSaU5EZ3pOREE1IiwidmVyIjoiaGFzaGVkcHJvb2Z0b2tlbiIsInNpdGVpZCI6Ik5UY3laalV5WW1NdE5EbGtPQzAwTVdaakxUaGhPVEV0TlRrNU5tUmhObU01TnpaaSIsImFwcF9kaXNwbGF5bmFtZSI6IlNoYXJlUG9pbnQgT25saW5lIENsaWVudCBFeHRlbnNpYmlsaXR5IFdlYiBBcHBsaWNhdGlvbiBQcmluY2lwYWwiLCJzaWduaW5fc3RhdGUiOiJbXCJrbXNpXCJdIiwiYXBwaWQiOiIyODQxMDU5Zi0wMTlmLTQ0YWYtODczMC04Mzc0NjA3YWQ5ZDkiLCJ0aWQiOiJkNGFmMmNiOC0wYjc4LTQ5YTQtYjViZC1jODJmZWE0MjBiYzgiLCJ1cG4iOiJhbGV4QGF0ZXJlbnRpZXYub25taWNyb3NvZnQuY29tIiwicHVpZCI6IjEwMDMyMDAwMzYxQjJDQTIiLCJjYWNoZWtleSI6IjBoLmZ8bWVtYmVyc2hpcHwxMDAzMjAwMDM2MWIyY2EyQGxpdmUuY29tIiwic2NwIjoibXlmaWxlcy5yZWFkIGdyb3VwLnJlYWQgZ3JvdXAud3JpdGUgYWxsc2l0ZXMucmVhZCBhbGxwcm9maWxlcy5yZWFkIiwidHQiOiIyIiwidXNlUGVyc2lzdGVudENvb2tpZSI6bnVsbH0.L0Fob1lFTUhiODZuNnRwTzhSZXppY3UzYnFIbytOUlhjTm5BVEU3NHN5Zz0%26version%3DPublished' (redirected from 'https://graph.microsoft.com/v1.0/sites/aterentiev.sharepoint.com,572f52bc-49d8-41fc-8a91-5996da6c976b,dcbb74db-3cec-47a9-a271-325a15b93034/lists/b8b00657-e9a7-4ff5-8609-f00571c0ca0e/items/1/driveItem/content?format=pdf') from origin 'null' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```
And I don't see any API to set `mode` to `no-cors` in `GraphRequest`

**Comment by patmill** [comment]
Hey @AJIXuMuK - tracking down some people to dig further into this.  Any chance you could get a fiddler trace of the various calls to graph / media service?  Feel free to send them to me, and I'll hand them off.

**Comment by AJIXuMuK** [comment]
Hey @patmill - sent to your email.

**Comment by bitflower** [comment]
Hey, getting the same `from origin 'null' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.` error. 

Do you have a workaround? Where can I set the headers in the Graph Client?

**Comment by bitflower** [comment]
Do we have the same problem if we use the node js API? I'm in the browser right now...

**Comment by Vishnuravi782** [comment]
any responses on this issue

**Comment by patmill** [comment]
Have you followed these instructions?  The flow is that you need to check the graph for a URL, but then you need to go and fetch the content from the URL returned.

https://docs.microsoft.com/en-us/graph/api/driveitem-get-content?view=graph-rest-1.0&tabs=http

**Comment by patmill** [comment]
Closing this issue, but if it still happens (following the guidance above), feel free to open a new issue.
